"""Command-line wrapper for fasttree."""
__docformat__ = "restructuredtext en"

from Bio.Application import _Option, _Switch, AbstractCommandline


class FastTreeCommandline(AbstractCommandline):
    """Command-line wrapper for FastTree.

    Homepage: http://www.microbesonline.org/fasttree/

    Citations:

    Price, M.N., Dehal, P.S., and Arkin, A.P. (2010) FastTree 2 -- Approximately 
    Maximum-Likelihood Trees for Large Alignments. PLoS ONE, 5(3):e9490. 
    doi:10.1371/journal.pone.0009490.

    """

    def __init__(self, cmd='fasttree', **kwargs):
        self.parameters = [                               
             _Option(['-out', 'output_input'],
                """The path to a Newick Tree output file needs to be specified first.  After that, using a separation of a space, 
                an input file of sequence alignments in fasta or phylip format is needed.  By default FastTree expects protein 
                alignments,  use -nt for nucleotides""",
                filename=False,
                is_required=True,
                equate=False,
                ),                            
            _Switch(['-nt', 'nt'],
                """ """,
                ),
            _Option(['-n', 'n'],
                """ """,
                ),
            _Switch(['-quote', 'quote'],
                """ """,
                ),
            _Switch(['-pseudo', '-pseudo 1.0', 'pseudo'],
                """ """,
                ),
            _Option(['-boot', 'boot'],
                """ """,
                ),
            _Switch(['-nosupport', 'nosupport'],
                """ """,
                ),         
            _Option(['-intree', 'intree'],
                """ """,
                ),                  
            _Option(['-intree1', 'intree1'],
                """ """,
                ),  
            _Switch(['-quiet', 'quiet'],
                """ """,
                ),  
            _Switch(['-nopr', 'nopr'],
                """ """,
                ),  
            _Option(['-nni', 'nni'],
                """ """,
                ),               
            _Option(['-spr', 'spr'],
                """ """,
                ),   
            _Switch(['-noml', 'noml'],
                """ """,
                ),       
            _Switch(['-mllen', 'mllen'],
                """ """,
                ),     
            _Option(['-mlnni', 'mlnni'],
                """ """,
                ),          
            _Option(['-mlacc', 'mlacc'],
                """ """,
                ),  
            _Option(['-cat', 'cat'],
                """ """,
                ),      
            _Switch(['-nocat', 'nocat'],
                """ """,
                ),
            _Switch(['-gamma', 'gamma'],
                """ """,
                ),
            _Switch(['-slow', 'slow'],
                """ """,
                ),
            _Switch(['-fastest', 'fastest'],
                """ """,
                ),     
            _Switch(['-2nd', 'Second'],
                """ """,
                ),
            _Switch(['-no2nd', 'no2nd'],
                """ """,
                ),
            _Switch(['-slownni', 'slownni'],
                """ """,
                ),  
            _Option(['-seed', 'seed'],
                """ """,
                ),      
            _Switch(['-top', 'top'],
                """ """,
                ),
            _Switch(['-notop', 'notop'],
                """ """,
                ),  
            _Option(['-topm', 'topm'],
                """ """,
                ),
            _Option(['-close', 'close'],
                """ """,
                ),
            _Option(['-refresh', 'refresh'],
                """ """,
                ),
            _Option(['-matrix', 'matrix'],
                """ """,
                ),
            _Switch(['-nomatrix', 'nomatrix'],
                """ """,
                ),
            _Switch(['-nj', 'nj'],
                """ """,
                ),
            _Switch(['-bionj', 'bionj'],
                """ """,
                ),     
            _Switch(['-wag', 'wag'],
                """ """,
                ),
            _Switch(['-gtr', 'gtr'],
                """ """,
                ),       
            _Option(['-gtrrates', 'gtrrates'],
                """ """,
                ),
            _Option(['-gtrfreq', 'gtrfreq'],
                """ """,
                ),
            _Option(['-constraints', 'constraints'],
                """ """,
                ),
            _Option(['-constraintWeight', 'constraintWeight'],
                """ """,
                ),          
            _Option(['-log', 'log'],
                """ """,
                ),      
            _Option(['-makematrix', 'makematrix'],
                """ """,
                ), 
            _Switch(['-rawdist', 'rawdist'],
                """ """,
                ),
            _Option(['-sprlength', 'sprlength'],
                """ """,
                ),      
            _Switch(['-nome', 'nome'],
                ###########################################################
                # Not sure if this should be a _Switch or _Option instead #  
                ###########################################################
                """ """,
                ),
                ]        

        AbstractCommandline.__init__(self, cmd, **kwargs)